src/main/java/org/reaktivity/nukleus/amqp/internal/stream/AmqpServerFactory.java

```diff
@@ -326,163 +983,392 @@ private void onNetwork(
+            while (progress <= limit && previous != decoder)
+            {
+                previous = decoder;
+                progress = decoder.decode(this, traceId, authorization, budgetId, buffer, progress, limit);
+            }
+
+            if (progress < limit)
+            {
+                if (decodeSlot == NO_SLOT)
                 {
-                    decodeState = this::decodeFrame;
+                    decodeSlot = bufferPool.acquire(initialId);
+                }
+
+                if (decodeSlot == NO_SLOT)
+                {
+                    cleanupNetwork(traceId, authorization);
                 }
                 else
                 {
-                    decodeState = this::decodeInvalid;
-                    doEnd();
+                    final MutableDirectBuffer slotBuffer = bufferPool.buffer(decodeSlot);
+                    decodeSlotLimit = limit - progress;
+                    slotBuffer.putBytes(0, buffer, progress, decodeSlotLimit);
                 }
             }
+            else
+            {
+                cleanupDecodeSlotIfNecessary();
+            }
+        }
+
+        private void onDecodeHeader(
+            long traceId,
+            long authorization,
+            AmqpProtocolHeaderFW header)
+        {
+            if (isAmqpHeaderValid(header))
+            {
+                doEncodeHeader(header.major(), header.minor(), header.revision(), traceId, authorization);
+            }
+        }
+
+        private void onDecodeFrameHeader(
+            AmqpFrameHeaderFW frameHeader)
+        {
+            doEncodeFrameHeader(frameHeader.doff(), frameHeader.type(), frameHeader.channel(), frameHeader.performative());
+        }
+
+        private void onDecodeOpen(
+            long traceId,
+            long authorization,
+            AmqpOpenFW open)
+        {
+            containerId = open.containerId().asString();
+            doEncodeOpen(traceId, authorization);
+        }
+
+        private void onDecodeBegin(
+            long traceId,
+            long authorization,
+            AmqpBeginFW begin)
+        {
+            if (begin.hasRemoteChannel() || sessions.containsKey(channelId))
+            {
+                onDecodeError(traceId, authorization, NOT_ALLOWED);
+            }
+            else
+            {
+                channelId++;
+                nextTransferId++;
+                AmqpSession session = sessions.computeIfAbsent(channelId, s -> new AmqpSession());
+                session.onDecodeBegin(traceId, authorization);
+            }
+        }
+
+        private void onDecodeAttach(
+            long traceId,
+            long authorization,
+            AmqpAttachFW attach)
+        {
+            if (!sessions.containsKey(channelId))
+            {
+                onDecodeError(traceId, authorization, NOT_ALLOWED);
+            }
+            else
+            {
+                sessions.get(channelId).onDecodeAttach(traceId, authorization, attach);
```

Comment:
Move `session.get(...)` to top, then test for `null` instead of doing both `contains` and `get` in the success case.